### ISSUE TITLE
Marking some fields in the returned object of a method NonNull and others Nullable?

### ISSUE THREAD

**Issue by fzyzcjy** [typetools/checker-framework#3296]
Hi! Thanks for the interesting library! A very simple question: I have a class (actually, a model, if you are interested) like - 

```
class Article {
    int id;
    String text;
    int authorId;
    Date createAt;
}
```

And I have a class returning it:

```
interface ArticleRepository {
    Article getFullArticle(); /* return the Article with ALL fields being non-null */ 
    Article getBriefArticle(); /* return the Article with only `id` and `text` fields being non-null */ 
}
```

Question: How can I annotate that properly? 

Thanks very much!

**Comment by mernst** [comment]
This is a good question.

The solution to your problem is qualifier polymorphism over classes.

Currently, the Checker Framework supports Java generics, and also qualifier polymorphims over methods.

Pull request #2883 implements qualifier polymorphims for classes.  We are hoping to complete that feature soon -- in the next release or two -- but this depends on how much time everyone has to work on it.

**Comment by fzyzcjy** [comment]
Thanks very much! Hoping to see that!

**Comment by wmdietl** [comment]
The limited form of class qualifier parameters in #2883 will not help in this situation - and it won't help with the Nullness Checker in general. The main qualifier of a type is reused for the qualifier parameter - but if the main qualifier is `@Nullable` one won't be able to access the field regardless.

In the `Article` example, you could try making the fields `@Nullable` and then having a method like this:

```java
    @EnsuresNonNullIf(result=true, expression="createAt")
    boolean isFullArticle() ...
```

We currently don't have a mechanism to specify as post-condition of `getFullArticle();` that the resulting `Article` will satisfy `isFullArticle`.

**Comment by fzyzcjy** [comment]
Thanks very much!

**Comment by mernst** [comment]
The [`@Unused` annotation](https://checkerframework.org/manual/#unused-fields) might already solve your problem.  It doesn't make a field nullable, but makes the field inaccessible, which is usually what you want.  You would need to create a pair of new annotations `@FullArticle` and `@BriefArticle`, then write a trivial type-checker that processes them.

**Comment by fzyzcjy** [comment]
Er sorry I do not quite get it... How do I write a "trivial type-checker", do you mean extending the existing Checker Framework?

**Comment by mernst** [comment]
Yes.  The process is explained at https://checkerframework.org/manual/#creating-a-checker .

**Comment by fzyzcjy** [comment]
Thanks very much!
